query-service-impl/src/main/java/org/hypertrace/core/query/service/postgres/QueryRequestToPostgresSQLConverter.java

```diff
@@ -0,0 +1,115 @@
+package org.hypertrace.core.query.service.postgres;
+
+import java.util.AbstractMap.SimpleEntry;
+import java.util.LinkedHashSet;
+import java.util.Map.Entry;
+import org.hypertrace.core.query.service.ExecutionContext;
+import org.hypertrace.core.query.service.api.Expression;
+import org.hypertrace.core.query.service.api.OrderByExpression;
+import org.hypertrace.core.query.service.api.QueryRequest;
+import org.hypertrace.core.query.service.api.SortOrder;
+import org.hypertrace.core.query.service.postgres.Params.Builder;
+import org.hypertrace.core.query.service.postgres.converters.ColumnRequestConverter;
+import org.hypertrace.core.query.service.postgres.converters.ColumnRequestConverterFactory;
+import org.hypertrace.core.query.service.postgres.converters.PostgresFunctionConverter;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Converts {@link QueryRequest} to Postgres SQL query */
+class QueryRequestToPostgresSQLConverter {
+
+  private static final Logger LOG =
+      LoggerFactory.getLogger(QueryRequestToPostgresSQLConverter.class);
+
+  private final TableDefinition tableDefinition;
+  private final PostgresFunctionConverter functionConverter;
+
+  QueryRequestToPostgresSQLConverter(
+      TableDefinition tableDefinition, PostgresFunctionConverter functionConverter) {
+    this.tableDefinition = tableDefinition;
+    this.functionConverter = functionConverter;
+  }
+
+  Entry<String, Params> toSQL(
+      ExecutionContext executionContext,
+      QueryRequest request,
+      LinkedHashSet<Expression> allSelections) {
+    ColumnRequestConverter columnRequestConverter =
+        ColumnRequestConverterFactory.getColumnRequestConverter(tableDefinition, functionConverter);
+    Builder paramsBuilder = Params.newBuilder();
+    StringBuilder pqlBuilder = new StringBuilder("Select ");
+    String delim = "";
+
+    // Set the DISTINCT keyword if the request has set distinctSelections.
+    if (request.getDistinctSelections()) {
+      pqlBuilder.append("DISTINCT ");
+    }
+
+    // allSelections contain all the various expressions in QueryRequest that we want selections on.
+    // Group bys, selections and aggregations in that order. See RequestAnalyzer#analyze() to see
+    // how it is created.
+    for (Expression expr : allSelections) {
+      pqlBuilder.append(delim);
+      pqlBuilder.append(
+          columnRequestConverter.convertSelectClause(expr, paramsBuilder, executionContext));
+      delim = ", ";
```

Comment:
nit: delim mgmt is a bit tricky, we can first collect all select clause transformations in the stream, and join at the end. we can do this as follow-up.